eclass/haskell-cabal: move cabal-build to end of cabal_src_compile.

ezzieyguywuf opened the pull request:
This "fixes" the `USE="doc"` test failures when a package has doctests.
The problem occurs when `./setup haddock` is run **before** `./setup build`

Signed-off-by: Wolfgang E. Sanyer <WolfgangESanyer@gmail.com>

This definitely doesn't appear to be be a gentoo-issue, but rather something going on in either `doctest` or `haddock` or both. I plan to dig into this a bit and hopefully file a bug report or PR upstream to resolve, but in the meantime this simple change shouldn't negatively impact things on our end...

ezzieyguywuf commented:
I will add, I can confirm that this change fixes the build failures in `machines-0.7.1` and `lens-4.19.1`

trofi commented:
At least historically haddock pass relied on already built `.hi` files to source interface information for doc generation. Reordering phases might have unexpected consequences to compiling package without optimisations (or even with wrong flags).

I'd like to understand why `doctests` are so special and fix doctests instead. Because otherwise other test suites seem to Just Work for some reason.

ezzieyguywuf commented:
@trofi fair point, let me try to figure out why this failure only occurs with doctests and not other tests.

As a side-note, at least in `lens` you can probably just mask `test-doctests` rather then disable the entire test suite.

ezzieyguywuf commented:
I've opened a [issue upstream](https://github.com/haskell/haddock/issues/1275), linking here for reference. Let's see if we get any traction there...

ezzieyguywuf commented:
@trofi , here's another stab at this, let me know your thoughts.

I've played around a bit in the `machines` build directory, and it seems that cabal (much like make) is smart enough to only re-build the parts that haddock messed up, so adding the second build command shouldn't be too "expensive" until we can figure out what's going on upstream.

trofi commented:
I'd still prefer to enable the hack only when explicitly requested by the ebuild.

ezzieyguywuf commented:
taking out of `[draft]` mode, I've added a new `FEATURES` flag to enable the hack, and I've updated two packages to show how this will fix the `USE=doc` test errors.

trofi commented:
> I've opened a [issue upstream](https://github.com/haskell/haddock/issues/1275), linking here for reference. Let's see if we get any traction there...

The bug is probably in `dev-haskell/doctest` or `dev-haskell/cabal-doctest`. `haddock` is just a tool that reads `.hi/.hs` files and writes out HTML. There is probably nothing to fix there.

trofi commented on eclass/haskell-cabal.eclass:
```diff
@@ -528,6 +528,10 @@ cabal_src_compile() {
+		# re-run build, to recover stuff that haddock makes disappear for some
+		# reason...thankfully, cabal is smart enough to only re-build the parts
+		# that have been lost, so this isn't too expensive.
+		cabal-build
```
The text should be more precise why we need it:
1. reference to a bug
2. symptoms of a failure

trofi commented on eclass/haskell-cabal.eclass:
```diff
@@ -16,19 +16,23 @@
-#                  only used for packages that use libghc internally and _must_
-#                  not pull upper versions
-#   test-suite --  add support for cabal test-suites (introduced in Cabal-1.8)
+#   haddock     --  for documentation generation
```
I would prefer not to reindent rest of options.

trofi commented on eclass/haskell-cabal.eclass:
```diff
@@ -16,19 +16,23 @@
+#                   only used for packages that use libghc internally and _must_
+#                   not pull upper versions
+#   test-suite  --  add support for cabal test-suites (introduced in Cabal-1.8)
+#   doctest-fix --  a hack that re-runs cabal-build after haddock documentation
```
I think the name should clearly say it's a workaround. `doctest-fix` sounds too specific. Maybe `rebuild-after-doc-workaround`?

The option description should explicitly list:
1. the symptom
2. have a link to bug description

trofi commented on eclass/haskell-cabal.eclass:
```diff
@@ -528,10 +533,12 @@ cabal_src_compile() {
-		# that have been lost, so this isn't too expensive.
-		cabal-build
+		if [[ -n "${CABAL_DOCTEST_FIX}" ]]; then
+			# re-run build, to recover stuff that haddock makes disappear for some
```
I also suggest adding `einfo` (or `ewarn`) here that describes enabled workaround.

trofi commented on eclass/haskell-cabal.eclass:
```diff
@@ -528,10 +533,12 @@ cabal_src_compile() {
-		# reason...thankfully, cabal is smart enough to only re-build the parts
-		# that have been lost, so this isn't too expensive.
-		cabal-build
+		if [[ -n "${CABAL_DOCTEST_FIX}" ]]; then
```
This change should probably be squashed with previous commit.

ezzieyguywuf commented:
> The bug is probably in `dev-haskell/doctest` or `dev-haskell/cabal-doctest`. `haddock` is just a tool that reads `.hi/.hs` files and writes out HTML. There is probably nothing to fix there.

I've filed a bug report with [the cabal project](https://github.com/haskell/cabal/issues/7213), I'll close out the haddock bug.

trofi commented:
I still suspect it's a `dev-haskell/doctest` or `dev-haskell/cabal-doctest` problem of a way it overrides default Cabal hooks and not a `dev-haskell/cabal` bug.

ezzieyguywuf commented:
I think dev-haskell/cabal-doctest is part of the cabal project, but I could
be wrong.

El sáb., 26 de diciembre de 2020 4:17 p. m., Sergei Trofimovich <
notifications@github.com> escribió:

> I still suspect it's a dev-haskell/doctest or dev-haskell/cabal-doctest
> problem of a way it overrides default Cabal hooks and not a
> dev-haskell/cabal bug.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gentoo-haskell/gentoo-haskell/pull/1115#issuecomment-751396993>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEMB5FIWZJRDWIOWI4VUIDSWZHGVANCNFSM4VHK3QCA>
> .
>

trofi commented:
It is a separate package: https://hackage.haskell.org/package/cabal-doctest

ezzieyguywuf commented:
> It is a separate package: https://hackage.haskell.org/package/cabal-doctest

This is what had me thinking it was wrapped up in cabal now https://github.com/haskellari/cabal-doctest/issues/59

Plus this https://github.com/haskell/cabal/tree/master/vendor/cabal-doctest-1.0.8

trofi commented:
Pulled. Thank you!
